### ISSUE TITLE
RKE fails to provision cluster with external etcd cluster

### ISSUE THREAD

**Issue by mristok** [rancher/rancher#15977]
**Rancher versions:**
rancher/rancher:stable

**Docker version: (`docker version`,`docker info` preferred)**
17.03.2-ce

**Operating system and kernel: (`cat /etc/os-release`, `uname -r` preferred)**
Ubuntu 16.04.4 LTS
4.4.0-120-generic

**Type/provider of hosts: (VirtualBox/Bare-metal/AWS/GCE/DO)**
VMware

**Setup details: (single node rancher vs. HA rancher, internal DB vs. external DB)**
HA Rancher

Related issue: https://github.com/rancher/rancher/issues/13097

**Steps to Reproduce:**
nodes:
  - address: <REDACTED>
    user: root
    role: [controlplane,etcd,worker]
    ssh_key_path: ~/.ssh/id_rsa
  - address: <REDACTED>
    user: root
    role: [controlplane,etcd,worker]
    ssh_key_path: ~/.ssh/id_rsa
  - address: <REDACTED>
    user: root
    role: [controlplane,etcd,worker]
    ssh_key_path: ~/.ssh/id_rsa
  - address: <REDACTED>
    user: root
    role: [controlplane,etcd,worker]
    ssh_key_path: ~/.ssh/id_rsa
  - address: <REDACTED>
    user: root
    role: [controlplane,etcd,worker]
    ssh_key_path: ~/.ssh/id_rsa

private_registries:
  - url: <PRIVATE REGISTRY>

network:
  plugin: canal

services:
  etcd:
    path: /etcdcluster
    external_urls:
      - https://<REDACTED>:2379
      - https://<REDACTED>:2379
      - https://<REDACTED>:2379
      - https://<REDACTED>:2379
      - https://<REDACTED>:2379
    ca_cert: |-
      -----BEGIN CERTIFICATE-----
      <INTERMEDIATE CA REDACTED>
      -----END CERTIFICATE-----
      -----BEGIN CERTIFICATE-----
      <ROOT CA REDACTED>
      -----END CERTIFICATE-----
    cert: |-
      -----BEGIN CERTIFICATE-----
      <REDACTED>
      -----END CERTIFICATE-----
    key: |-
      -----BEGIN PRIVATE KEY-----
      <REDACTED>
      -----END PRIVATE KEY-----

system_images:
  etcd: <PRIVATE REGISTRY>/rancher/coreos-etcd:v3.2.24
  alpine: <PRIVATE REGISTRY>/rancher/rke-tools:v0.1.15
  nginx_proxy: <PRIVATE REGISTRY>/rancher/rke-tools:v0.1.15
  cert_downloader: <PRIVATE REGISTRY>/rancher/rke-tools:v0.1.15
  kubernetes_services_sidecar: <PRIVATE REGISTRY>/rancher/rke-tools:v0.1.15
  kubedns: <PRIVATE REGISTRY>/rancher/k8s-dns-kube-dns-amd64:1.14.13
  dnsmasq: <PRIVATE REGISTRY><PRIVATE REGISTRY>/rancher/k8s-dns-dnsmasq-nanny-amd64:1.14.13
  kubedns_sidecar: <PRIVATE REGISTRY>/rancher/k8s-dns-sidecar-amd64:1.14.13
  kubedns_autoscaler: <PRIVATE REGISTRY>/rancher/cluster-proportional-autoscaler-amd64:1.0.0
  kubernetes: <PRIVATE REGISTRY>/rancher/hyperkube:v1.10.3-rancher2
  flannel: <PRIVATE REGISTRY>/rancher/coreos-flannel:v0.10.0
  flannel_cni: <PRIVATE REGISTRY>/rancher/coreos-flannel-cni:v0.3.0
  calico_node: <PRIVATE REGISTRY>/rancher/calico-node:v3.1.3
  calico_cni: <PRIVATE REGISTRY>/rancher/calico-cni:v3.1.3
  calico_ctl: <PRIVATE REGISTRY>/rancher/calico-ctl:v2.0.0
  canal_node: <PRIVATE REGISTRY>/rancher/calico-node:v3.1.3
  canal_cni: <PRIVATE REGISTRY>/rancher/calico-cni:v3.1.3
  canal_flannel: <PRIVATE REGISTRY>/rancher/coreos-flannel:v0.10.0
  pod_infra_container: <PRIVATE REGISTRY>/rancher/pause-amd64:3.1
  metrics_server: <PRIVATE REGISTRY>/rancher/metrics-server-amd64:v0.3.1

addons: |-
  ---
  kind: Namespace
  apiVersion: v1
  metadata:
    name: cattle-system
  ---
  kind: ServiceAccount
  apiVersion: v1
  metadata:
    name: cattle-admin
    namespace: cattle-system
  ---
  kind: ClusterRoleBinding
  apiVersion: rbac.authorization.k8s.io/v1
  metadata:
    name: cattle-crb
    namespace: cattle-system
  subjects:
  - kind: ServiceAccount
    name: cattle-admin
    namespace: cattle-system
  roleRef:
    kind: ClusterRole
    name: cluster-admin
    apiGroup: rbac.authorization.k8s.io
  ---
  apiVersion: v1
  kind: Secret
  metadata:
    name: cattle-keys-ingress
    namespace: cattle-system
  type: Opaque
  data:
    tls.crt: <REDACTED>
    tls.key: <REDACTED>
  ---
  apiVersion: v1
  kind: Secret
  metadata:
    name: cattle-keys-server
    namespace: cattle-system
  type: Opaque
  data:
    cacerts.pem: <REDACTED>
  ---
  apiVersion: v1
  kind: Service
  metadata:
    namespace: cattle-system
    name: cattle-service
    labels:
      app: cattle
  spec:
    ports:
    - port: 80
      targetPort: 80
      protocol: TCP
      name: http
    - port: 443
      targetPort: 443
      protocol: TCP
      name: https
    selector:
      app: cattle
  ---
  apiVersion: extensions/v1beta1
  kind: Ingress
  metadata:
    namespace: cattle-system
    name: cattle-ingress-http
    annotations:
      nginx.ingress.kubernetes.io/proxy-connect-timeout: "30"
      nginx.ingress.kubernetes.io/proxy-read-timeout: "1800"
      nginx.ingress.kubernetes.io/proxy-send-timeout: "1800"
      nginx.ingress.kubernetes.io/affinity: "cookie"
      nginx.ingress.kubernetes.io/session-cookie-name: "route"
      nginx.ingress.kubernetes.io/session-cookie-hash: "sha1"
  spec:
    rules:
    - host: rancher.local
      http:
        paths:
        - backend:
            serviceName: cattle-service
            servicePort: 80
    tls:
    - secretName: cattle-keys-ingress
      hosts:
      - rancher.local
  ---
  kind: Deployment
  apiVersion: extensions/v1beta1
  metadata:
    namespace: cattle-system
    name: cattle
  spec:
    replicas: 3
    template:
      metadata:
        labels:
          app: cattle
      spec:
        serviceAccountName: cattle-admin
        containers:
        - image: <PRIVATE REGISTRY>/rancher/rancher:stable
          imagePullPolicy: Always
          name: cattle-server
          env:
          - name: CATTLE_SYSTEM_DEFAULT_REGISTRY
            value: <PRIVATE REGISTRY>
          - name: http_proxy
            value: http://<REDACTED>:3128
          - name: https_proxy
            value: http://<REDACTED>:3128
          - name: no_proxy
            value: localhost,127.0.0.1,10.0.0.0/8,172.16.0.0/12,192.168.0.0/16
          - name: NO_PROXY
            value: localhost,127.0.0.1,10.0.0.0/8,172.16.0.0/12,192.168.0.0/16
          ports:
          - containerPort: 80
            protocol: TCP
          - containerPort: 443
            protocol: TCP
          volumeMounts:
          - mountPath: /etc/rancher/ssl
            name: cattle-keys-volume
            readOnly: true
        volumes:
        - name: cattle-keys-volume
          secret:
            defaultMode: 420
            secretName: cattle-keys-server

**Results:**
Failure during RKE deployment:
FATA[0075] [controlPlane] Failed to bring up Control Plane: Failed to verify healthcheck: Service [kube-apiserver] is not healthy on host [10.0.14.157]. Response code: [403], response body: {"kind":"Status","apiVersion":"v1","metadata":{},"status":"Failure","message":"forbidden: User \"kube-apiserver\" cannot get path \"/healthz\"","reason":"Forbidden","details":{},"code":403}
, log: I1006 04:17:38.031751       1 naming_controller.go:276] Starting NamingConditionController

Additionally, it appears that the external ETCD Private Key is not properly being mounted to the kube-apiservers.

**Comment by mristok** [comment]
@loganhz Any updates on this. It has halted our confidence in pursuing and deploying Rancher into production.

**Comment by alena1108** [comment]
@galal-hussein can you check please

**Comment by galal-hussein** [comment]
@mristok the clsuter.yml is not correct, using external etcd will conflict with you specifying nodes with role etcd, you can only use either of options:
1- nodes with roles etcd (this will make rke generates certs for each etcd nodes)
2- external etcd (this will require you to point to the external etcd node and specify the certs)

**Comment by mristok** [comment]
@galal-hussein apparently the yaml I submitted was out of date, because I corrected that error (WRT etcd as a role) during my build out. The system is isolated and behind a corporate proxy (not fully air gapped). The above was failing using corporate PKI certs, but works with self-signed cfssl certs for the external etcd cluster. The etcd certs are specified under services.etcd

A major issue that I repeatedly experience is that the etcd certs are passed to the docker hosts, but the private-key is always invalid. If I replace the private-key after initial failure and re-run rke up, it will successfully connect to the external etcd and finish the k8s build.
